File: python/day3.py
import collections
import re

INPUT = '../inputs/day3.txt'
REGEX = r'#(?P<id>\d+)\s@\s(?P<x>\d+),(?P<y>\d+):\s(?P<w>\d+)x(?P<h>\d+)'

Point = collections.namedtuple('Point', 'x y')
Rectangle = collections.namedtuple('Rectangle', 'p1 p2')
Claim = collections.namedtuple('Claim', 'id bounds width height')


def part1(claims):
    for index, claim in enumerate(claims):
        for comparison in claims[index + 1:]:
            intersection = calculate_intersection(
                claim.bounds, comparison.bounds)
            isvalid = isvalid_rectangle(intersection)


def calculate_intersection(rectangle1, rectangle2):
    x1 = max(rectangle1.p1.x, rectangle2.p1.x)
    y1 = max(rectangle1.p1.y, rectangle2.p1.y)
    x2 = max(rectangle1.p2.x, rectangle2.p2.x)
    y2 = max(rectangle1.p2.y, rectangle2.p2.y)
    return Rectangle(Point(x1, y1), Point(x2, y2))


def isvalid_rectangle(rectangle):
    # TODO: Actually check if rectangle is valid
    return True


def parsefile(path, regex):
    with open(path, mode='r') as data:
        return [parseline(l, regex) for l in data.readlines()]


def parseline(line, regex):
    match = re.match(regex, line)
    groups = match.groupdict()
    x1 = int(groups['x'])
    y1 = int(groups['y'])
    w = int(groups['w'])
    h = int(groups['h'])
    x2 = x1 + w
    y2 = y1 + h
    return Claim(int(groups['id']), Rectangle(Point(x1, y1), Point(x2, y2)), w, h)


if __name__ == '__main__':
    claims = parsefile(INPUT, REGEX)
    part1_test_data = [
        Claim(1, Rectangle(Point(1, 3), Point(5, 8)), 4, 4),
        Claim(2, Rectangle(Point(3, 1), Point(8, 5)), 4, 4),
        Claim(3, Rectangle(Point(5, 5), Point(7, 7)), 2, 2)
    ]
    part1(claims)

File: elixir/day3.exs
defmodule Point do
    @enforce_keys [:x, :y]
    defstruct [:x, :y] 
end

defmodule Rectangle do
    @enforce_keys [:p1, :p2]
    defstruct [:p1, :p2] 
end

defmodule Claim do
    @enforce_keys [:id, :bounds]
    defstruct [:id, :bounds] 
end

defmodule Day3 do
    def part1(path, regex) do
        # compiled_regex = Regex.compile!(regex)
        lines = path
            |> File.read!()
            |> String.split("\r\n", trim: true)
            |> Enum.map(fn l -> parse_line(l, regex) end)

        lines
        |> Enum.with_index()
        |> Enum.reduce(MapSet.new(), fn {x, i}, acc ->
            points = lines
                |> Enum.drop(i + 1)
                |> Enum.map(fn y -> calculate_intersection(x.bounds, y.bounds) end)
                |> Enum.filter(fn i -> !!i end)
                |> Enum.map(fn i -> get_all_points(i) end)
                |> List.flatten()
            MapSet.union(acc, MapSet.new(points))
        end)
        |> MapSet.size()
        |> IO.inspect()
    end

    defp parse_line(line, regex) when is_binary(line) do
        matches = Regex.named_captures(regex, line)
        |> Enum.map(fn {k, v} -> {String.to_atom(k), String.to_integer(v)} end)
        |> Enum.into(%{})

        p1 = %Point{x: matches.x, y: matches.y}
        p2 = %Point{x: matches.x + matches.w - 1, y: matches.y + matches.h - 1}
        %Claim{id: matches.id, bounds: %Rectangle{p1: p1, p2: p2}}
    end

    defp calculate_intersection(rectangle1, rectangle2) when is_map(rectangle1) and is_map(rectangle2) do
        x1 = Enum.max([rectangle1.p1.x, rectangle2.p1.x])
        y1 = Enum.max([rectangle1.p1.y, rectangle2.p1.y])
        x2 = Enum.min([rectangle1.p2.x, rectangle2.p2.x])
        y2 = Enum.min([rectangle1.p2.y, rectangle2.p2.y])
        is_valid = x1 <= x2 and y1 <= y2
        if is_valid do
            p1 = %Point{x: x1, y: y1}
            p2 = %Point{x: x2, y: y2}
            %Rectangle{p1: p1, p2: p2}
        else
            nil
        end
    end

    defp get_all_points(rectangle) when is_map(rectangle) do
        x_range = rectangle.p1.x..rectangle.p2.x
        y_range = rectangle.p1.y..rectangle.p2.y

        x_range
        |> Enum.map(fn x -> Enum.map(y_range, fn y -> {x, y} end) end)
        |> List.flatten()
    end
end
regex = ~r/#(?<id>\d+)\s@\s(?<x>\d+),(?<y>\d+):\s(?<w>\d+)x(?<h>\d+)/
test_path = "../inputs/day3_test.txt"
Day3.part1(test_path, regex)

path = "../inputs/day3.txt"
Day3.part1(path, regex)